src/core/classes/language.ts

```diff
@@ -0,0 +1,96 @@
+import { extend, cloneGrammar, resolveGrammar } from '../../shared/language-util';
+import type LanguageRegistry from './language-registry';
+import type { ComponentProtoBase } from './language-registry';
+import type { Grammar, GrammarOptions } from '../../grammar';
+
+export default class Language {
+	def: LanguageProto;
+	registry: LanguageRegistry;
+	evaluatedGrammar?: Grammar;
+
+	constructor (def: LanguageProto, registry: LanguageRegistry) {
+		this.def = def;
+		this.registry = registry;
+	}
+
+	get id () {
+		return this.def.id;
+	}
+
+	get alias () {
+		if (!this.def.alias) {
+			return [];
+		}
+
+		return Array.isArray(this.def.alias) ? this.def.alias : [this.def.alias];
+	}
+
+	get base (): Language | null {
+		if (!this.def.base) {
+			return null;
+		}
+
+		return this.registry.get(this.def.base.id) ?? null;
+	}
+
+	get extends () {
+		if (!this.def.extends) {
+			return [];
+		}
+		return Array.isArray(this.def.extends) ? this.def.extends : [this.def.extends];
+	}
+
+	get grammar (): Grammar {
+		// Lazily evaluate grammar
+		const def = this.def;
+
+		let { id, grammar } = def;
+		let base = this.base;
+
+		if (typeof grammar === 'function') {
+			grammar = grammar.call(this, {
+				base,
+				getLanguage (id: string) {
+					return this.registry.get(id);
+				},
```

Comment:
I wonder if this should be an error function? Will `registry` be accessible if not?

Reply:
Oh, yes it does need to be an arrow function! Feel free to commit directly to this PR